llama-index-core/llama_index/core/query_engine/sub_question_query_engine.py

```diff
@@ -129,83 +148,53 @@ def from_defaults(
-            colors = get_color_mapping([str(i) for i in range(len(sub_questions))])
-
-            if self._verbose:
-                print_text(f"Generated {len(sub_questions)} sub questions.\n")
-
-            if self._use_async:
-                tasks = [
-                    self._aquery_subq(sub_q, color=colors[str(ind)])
-                    for ind, sub_q in enumerate(sub_questions)
-                ]
-
-                qa_pairs_all = run_async_tasks(tasks)
-                qa_pairs_all = cast(List[Optional[SubQuestionAnswerPair]], qa_pairs_all)
-            else:
-                qa_pairs_all = [
-                    self._query_subq(sub_q, color=colors[str(ind)])
-                    for ind, sub_q in enumerate(sub_questions)
-                ]
-
-            # filter out sub questions that failed
-            qa_pairs: List[SubQuestionAnswerPair] = list(filter(None, qa_pairs_all))
-
-            nodes = [self._construct_node(pair) for pair in qa_pairs]
-
-            source_nodes = [node for qa_pair in qa_pairs for node in qa_pair.sources]
-            response = self._response_synthesizer.synthesize(
-                query=query_bundle,
-                nodes=nodes,
-                additional_source_nodes=source_nodes,
-            )
-
-            query_event.on_end(payload={EventPayload.RESPONSE: response})
-
-        return response
-
-    async def _aquery(self, query_bundle: QueryBundle) -> RESPONSE_TYPE:
-        with self.callback_manager.event(
-            CBEventType.QUERY, payload={EventPayload.QUERY_STR: query_bundle.query_str}
-        ) as query_event:
-            sub_questions = await self._question_gen.agenerate(
-                self._metadatas, query_bundle
-            )
-
-            colors = get_color_mapping([str(i) for i in range(len(sub_questions))])
-
-            if self._verbose:
-                print_text(f"Generated {len(sub_questions)} sub questions.\n")
-
-            tasks = [
-                self._aquery_subq(sub_q, color=colors[str(ind)])
-                for ind, sub_q in enumerate(sub_questions)
-            ]
+    @step
+    async def generate_sub_questions(
+        self, ctx: Context, ev: StartEvent
+    ) -> SubQuestionEvent | None:
+        """Generate sub questions."""
+        sub_questions = await self._question_gen.agenerate(
+            self._metadatas, ev.query_bundle
+        )
+        colors = get_color_mapping([str(i) for i in range(len(sub_questions))])
 
-            qa_pairs_all = await asyncio.gather(*tasks)
-            qa_pairs_all = cast(List[Optional[SubQuestionAnswerPair]], qa_pairs_all)
+        ctx.data["num_sub_questions"] = len(sub_questions)
+        ctx.data["query_bundle"] = ev.query_bundle
 
-            # filter out sub questions that failed
-            qa_pairs: List[SubQuestionAnswerPair] = list(filter(None, qa_pairs_all))
+        for ind, sub_q in enumerate(sub_questions):
+            self.send_event(SubQuestionEvent(query=sub_q, color=colors[str(ind)]))
```

Comment:
I think we already merged the PR that moves `send_event()` to be on the context rather than the `self` instance